src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,397 @@
+import java.net.URL;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.*;
+
+public class S3SignerV4 extends S3Signer {
+    private static final String HEADER_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss Z";
+    //The timestamp must be in UTC and in the following ISO 8601 format: YYYYMMDD'T'HHMMSS'Z'
+    private static final String AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String AMZ_DATE_FORMAT_SHORT = "yyyyMMdd";
+    private static final long PRESIGN_URL_MAX_EXPIRATION_SECONDS = 60 * 60 * 24 * 7;
+    private static final SortedSet<String> excludedSignedHeaders = new TreeSet(Arrays.asList(
+        "authorization"
+    ));
+
+    public S3SignerV4(S3Config s3Config) {
+        super(s3Config);
+    }
+
+    public void sign(ClientRequest request, String resource, Map<String, String> parameters, Map<String, List<Object>> headers) {
+        // # Preparation, add x-amz-date and host headers
+        String date = null;
+        String serviceType = getServiceType();
+        if (headers.containsKey(S3Constants.AMZ_DATE)) {
+            date = RestUtil.getFirstAsString(headers, S3Constants.AMZ_DATE);
+        }
+        else {
+            date = getDate(parameters, headers);
+        }
+        String shortDate = getShortDate(date);
+        addHeadersForV4(request.getURI(), date, headers);
+
+        // #1 Create a canonical request for Signature Version 4
+        String canonicalRequest = getCanonicalRequest(request, parameters, headers);
+
+        // #2 Create a string to sign for Signature Version 4
+        String stringToSign = getStringToSign(request.getMethod(), resource, parameters, headers, date, serviceType, canonicalRequest);
+        log.debug("StringToSign: {}", stringToSign);
+
+        // Get signed headers
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+
+        // #3 Calculate the signature for AWS Signature Version 4
+        byte[] key = getSigningKey(shortDate, serviceType);
+        String signature = getSignature(stringToSign, key);
+        log.debug("Signature: {}", signature);
+
+        // #4 Adding signing information to the authorization header
+        RestUtil.putSingle(headers, "Authorization", S3Constants.AWS_HMAC_SHA256_ALGORITHM +
+                " Credential=" + s3Config.getIdentity() + "/" + shortDate +
+                "/" + S3Constants.AWS_DEFAULT_REGION + "/" + serviceType + "/" + S3Constants.AWS_V4_TERMINATOR +
+                ", SignedHeaders=" + signedHeaders.toString() + ", " + S3Constants.PARAM_SIGNATURE + "= " + signature);
```

Comment:
there is a space after the equals here - is that going to cause any problems?

Reply:
Removed the space.